### DISCUSSION TITLE
Just Download the new application file.

### DISCUSSION THREAD

**Discussion by ritishbhardwaj** [dennisvang/tufup#142] [Q&A]
I want to just download the new .exe file from the server and do not want to overwrite it on the the place of older .exe .
The function ```download_and_apply_update``` is also updating the older .exe.
Can you explain me how can i do it gracefully
Also i tried ```get_targetinfo(target_path: str | [TargetMeta])→ TargetFile | None``` but unable to understand what is this TargetMeta and from where to get it....

**Comment by ritishbhardwaj** [comment]
I have tried some experiments, like 
```
            target_filename = any_update.target_path_str

            target_path = os.path.join(meta_dir, target_filename)
            print(f"Target path to get info: {target_path}")

            mytrinfo= client.get_targetinfo(target_path=target_filename)
            test_return=  client.download_target(targetinfo=mytrinfo)
```

and this is successfully downloading, but the file which is getting downloading is .tar.gz file, which i don't want.
 I want the functionality within the tufup which , same like ```download_and_apply_updates()```, will extract the .exe file and placed in ```settings.INSTALL_DIR``` and then remove the .tar.gz from the targets directory of the Client side repo directory.

**Reply by dennisvang** [reply]
>... and this is successfully downloading, but the file which is getting downloading is .tar.gz file, which i don't want. ...

The `.tar.gz` archive is extracted to a temporary directory in the [`Client._apply_updates()`][1] step, right before calling the `install` function mentioned in my [answer to your previous question][2].

To change what happens with the extracted files, you have two options:

- specify a custom `install` function: Anything you want to do with the extracted archive can be specified in a custom `install` function, which can be passed into [`Client.download_and_apply_update()`][3], as explained in the [example here][7]. The install function is platform dependent, so, as a starting point for a custom `install` function on Windows, you could copy the default [`_install_update_win()`][4] and modify that as necessary.

- specify a custom `batch_template`: Depending on your specific needs, you could specify a custom Windows [batch template][6], as explained in the [example here][5] (either *in stead of*, or *in addition to* a custom install function).
 
Basically, on Windows, the *default* `install` function opens a new process to run the steps from the batch template, then closes the app so that its files can be replaced. The *default* batch template does a copy/overwrite using robocopy. You can modify this to meet your needs.
 
Note that much of the complexity of the windows install function and batch template comes from the fact that windows does not allow [overwriting a running executable][9]. However, if you are *not* overwriting the *running* .exe, then you may *not* need to use a separate process with a batch template at all. In that case you could just specify a highly simplified `install` function very similar to the [Mac version][8].

[1]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/client.py#L315
[2]: https://github.com/dennisvang/tufup/discussions/140#discussioncomment-9591216
[3]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/client.py#L99
[4]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/utils/platform_specific.py#L126
[5]: https://github.com/dennisvang/tufup-example/issues/5#issuecomment-1262195266
[6]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/utils/platform_specific.py#L88
[7]: https://github.com/dennisvang/tufup-example/discussions/17#discussioncomment-7100228
[8]: https://github.com/dennisvang/tufup/blob/4bb16ad228b4b6f39e42527cc4367e501d728142/src/tufup/utils/platform_specific.py#L237
[9]: https://stackoverflow.com/q/6695496
